gax/src/test/java/com/google/api/gax/batching/v2/FlowControllerTest.java

```diff
@@ -0,0 +1,253 @@
+                      permitsReserved.set(null);
+                      flowController.reserve(maxNumBytes);
+                    } catch (FlowControlException e) {
+                      throw new AssertionError(e);
+                    }
+                  }
+                });
+
+    permitsReserved.get();
+    flowController.release(1, 1);
+    finished.get();
+  }
+
+  @Test
+  public void testReserveRelease_rejectedByElementCount_noBytesLimit() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(1)
+                .setLimitExceededBehavior(LimitExceededBehavior.ThrowException)
+                .build());
+
+    testRejectedReserveRelease(flowController, 5, MaxOutstandingElementCountReachedException.class);
+  }
+
+  @Test
+  public void testReserveRelease_rejectedByNumberOfBytes() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(100)
+                .setMaxOutstandingRequestBytes(10L)
+                .setLimitExceededBehavior(LimitExceededBehavior.ThrowException)
+                .build());
+
+    testRejectedReserveRelease(
+        flowController, 10, MaxOutstandingRequestBytesReachedException.class);
+  }
+
+  @Test
+  public void testReserveRelease_rejectedByNumberOfBytes_noElementCountLimit() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingRequestBytes(10L)
+                .setLimitExceededBehavior(LimitExceededBehavior.ThrowException)
+                .build());
+
+    testRejectedReserveRelease(
+        flowController, 10, MaxOutstandingRequestBytesReachedException.class);
+  }
+
+  @Test
+  public void testRestoreAfterFail() throws FlowControlException {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(2)
+                .setMaxOutstandingRequestBytes(5L)
+                .setLimitExceededBehavior(LimitExceededBehavior.ThrowException)
+                .build());
+
+    flowController.reserve(1);
+
+    try {
+      flowController.reserve(5);
+      fail("flowController should not have any bytes left");
+    } catch (MaxOutstandingRequestBytesReachedException e) {
+      assertThat(e.getCurrentMaxBatchBytes()).isEqualTo(5);
+    }
+
+    flowController.release(2, 5);
+  }
+
+  private void testRejectedReserveRelease(
+      FlowController flowController,
+      int maxNumBytes,
+      Class<? extends FlowControlException> expectedException)
+      throws FlowControlException {
+
+    flowController.reserve(1);
+
+    try {
+      flowController.reserve(maxNumBytes);
+      fail("Should thrown a FlowController.FlowControlException");
```

Comment:
Please use Truth consistently

Reply:
Thanks for the suggestion, I have removed Assert#fail occurrences.